File: 백준/Silver/2667. 단지번호붙이기/단지번호붙이기.py
def dfs(a, b):
    graph[a][b] = 0
    global house_cnt
    house_cnt += 1
    if b+1 < n and graph[a][b+1] == 1:
        dfs(a, b+1)
    if b-1 >= 0 and graph[a][b-1] == 1:
        dfs(a, b-1)
    if a+1 < n and graph[a+1][b] == 1:
        dfs(a+1, b)
    if a-1 >= 0 and graph[a-1][b] == 1:
        dfs(a-1, b)


n = int(input())
graph = []
for i in range(n):
    graph.append(list(map(int, input())))
cnt = 0
house_cnt = 0
house = []
for i in range(n):
    for j in range(n):
        if graph[i][j] == 1:
            dfs(i, j)
            cnt += 1
            house.append(house_cnt)
            house_cnt = 0
print(cnt)
house.sort()
for i in house:
    print(i)

File: 백준/Silver/2667. 단지번호붙이기/단지번호붙이기.cc
#include <iostream>
#include <vector>
#include <algorithm>
using namespace std;

int N;
int houseCount = 0;
vector<string> map;
bool isVisited[25][25] = {false};
int dx[4] = {1, -1, 0, 0};
int dy[4] = {0, 0, 1, -1};

void dfs(int x, int y)
{
    isVisited[x][y] = true;
    houseCount++;
    for (int i = 0; i < 4; i++)
    {
        int nextX = x + dx[i];
        int nextY = y + dy[i];
        if (nextX >= 0 && nextX < N && nextY >= 0 && nextY < N)
        {
            if (!isVisited[nextX][nextY] && map[nextX][nextY] == '1')
                dfs(nextX, nextY);
        }
    }
}

int main()
{
    int result = 0;
    cin >> N;
    vector<int> house;
    for (int i = 0; i < N; i++)
    {
        string temp;
        cin >> temp;
        map.push_back(temp);
    }
    for (int i = 0; i < N; i++)
    {
        for (int j = 0; j < N; j++)
        {
            if (map[i][j] == '1' && !isVisited[i][j])
            {
                dfs(i, j);
                result++;
                house.push_back(houseCount);
                houseCount = 0;
            }
        }
    }
    sort(house.begin(), house.end());
    cout << result << "\n";
    for (int i = 0; i < house.size(); i++)
        cout << house[i] << "\n";
    return 0;
}